design/proposal-concurrent-collectors.md

```diff
@@ -0,0 +1,55 @@
+# Run collectors concurrently
+ 
+## Goals
+
+Increase the speed at which collectors complete.
+
+Ensure that the cluster-resources collector runs prior to any others, so that things like pod status and pod logs only include the pods from the application rather than the ones that Troubleshoot started as part of collectors.
+## Non Goals
+
+
+## Background
+
+Currently all collectors run sequentially, in the order in which they are specified in the spec.  This is implied, and has not been guaranteed.  With the addition of the ability to supply multiple specs, the order is no longer predicatble.
+
+Although sequence is not guaranteed for any collectors, #768 ensures that the `clusterResources` collector runs first, so that the list of resources does not include those that are started during the Troubleshoot collection sequence.
```

Comment:
There has traditionally been usage of collectors in the past that relied on them being sequential. Do we intend scrap those use cases with this change? An example:

```
    - exec:
        name: get-latest-prometheus-snapshot
        collectorName: get-latest-prometheus-snapshot
        namespace: some-namespace
        selector:
          - app=some-app
        command: ["sh"]
        args: ["-c", "cd /prometheus/data/snapshots && mv `ls -t | head -n 1` latest"]
        timeout: 5s
    - copy:
        name: collect-prometheus-snapshot
        collectorName: collect-prometheus-snapshot
        selector:
          - app=some-app
        namespace: some-namespace
        containerPath: /prometheus/data/snapshots/latest
```

The copy collector relies on the exec collector preceding it. To continue supporting this after collectors run concurrently, we'd likely need some sort of `depends_on` functionality

Reply:
This is an interesting use case, I'm curious if it's the correct way to encourage people to do this sort of operation. I think a depends_on might be useful, but I'm also concerned that the example given above isn't idempotent. Each run creates and replaces the 'latest' file in the pod and doesn't clean it up when it's done!

For this example, would it be better to have a collector that gathers that latest snapshot w/o creating artifacts in the pod and returns that? I'm wondering if the use case for dependent collectors is the best way to address issues or if it's an indication we could use some more/better collectors?

Reply:
There's no guarantee of sequencing, and even less so when we have multiple specs merged.  The more we merge specs, the less we can predict the sequence.

I'm keener, for this use case and others like it, to write a new collector that would, e.g. run a command and process the result somehow.  For this example, would we not pipe the result to stdout which is collected in any case?

I did write up a proposal for a depends-on type of field but that is a complex subject which would need some extra business justification for the effort involved in writing and maintaining it.

Reply:
I think the depends-on feature would need to be a separate design doc, it's not something we have right now and only works currently by accident.  The fact that we now merge specs makes it less predictable and I'd discourage anyone from relying on the order of collectors.